### ISSUE TITLE
Mentioning someone in message

### ISSUE THREAD

**Issue by AlexandarY** [jenkinsci/office-365-connector-plugin#55]
Hey,
I was wondering if you could add an option, where when the message is posted in the MS Teams channel, the person who started the build or the author of the last GIT_COMMIT could be mentioned?
Thanks!

**Comment by damianszczepanik** [comment]
It's working now as you described

**Comment by casz** [comment]
webhooks on teams does not support personal mentions.
Microsoft has implemented follow channel and connectors can then notify you.
https://microsoftteams.uservoice.com/forums/555103-public/suggestions/17022934-notification-for-connectors
![image](https://user-images.githubusercontent.com/1661688/35650830-b6ca9480-06dd-11e8-9e10-b1f0b63ce451.png)

**Comment by damianszczepanik** [comment]
![remark](https://user-images.githubusercontent.com/9612911/35707646-e4f45f26-07aa-11e8-8eaa-f47e7ff83adf.png)
I have information who has started the build. Similar is for the author of the change that triggers the build

**Comment by casz** [comment]
Your missing the points of teams mention, he wants to @randomuser to sent notification to a specific user.

**Comment by AlexandarY** [comment]
As per @casz, it seems indeed the Cards do not support mention. The use case was in case the build failed, I could use Culprits/Developer to mention them so that if they do not follow the channel, I could get their attention.
Maybe in the future, if the @mention feature is added to the Cards, we can look into this again?

**Comment by HenrikPilz** [comment]
Is there any possibility for this feature to come in near future?

**Comment by casz** [comment]
That is is question you should direct at microsoft @HenrikPilz

**Comment by damianszczepanik** [comment]
I haven't seen support for that

**Comment by tayyab23** [comment]
@damianszczepanik 
Wondering if this can be re-opened in presence of [this doc](https://docs.microsoft.com/en-us/microsoftteams/platform/task-modules-and-cards/cards/cards-format?tabs=adaptive-md%2Cconnector-html#mention-support-within-adaptive-cards)

I haven't been able to make it work using the existing `factDefinitions` using the <at>John Doe</at>

**Comment by damianszczepanik** [comment]
I guess so

**Comment by miclefebvre** [comment]
@damianszczepanik is it something that would be available in the near future ? We are migrating from Slack and it's a show stopper for us. We would have to go back to email if not possible. 

Also is there a way add some documentation on what factDefinitions is ? It's mentioned in the doc, but I have no idea what it's supposed to do.

**Comment by tayyab23** [comment]
> @damianszczepanik is it something that would be available in the near future ? We are migrating from Slack and it's a show stopper for us. We would have to go back to email if not possible.
> 
> Also is there a way add some documentation on what factDefinitions is ? It's mentioned in the doc, but I have no idea what it's supposed to do.

@miclefebvre 
I specified items from Status to CTL Url in fact definitions, the card in the notification appears like so
![image](https://user-images.githubusercontent.com/35904881/97749325-1b522f00-1aac-11eb-9eec-6abad24894d3.png)

**Comment by damianszczepanik** [comment]
I would be happy to accept PR for this and help but not sure when have time for developing this feature from scratch.

Anyway this looks like valuable change!

**Comment by leonardobsjr** [comment]
> > @damianszczepanik is it something that would be available in the near future ? We are migrating from Slack and it's a show stopper for us. We would have to go back to email if not possible.
> > Also is there a way add some documentation on what factDefinitions is ? It's mentioned in the doc, but I have no idea what it's supposed to do.
> 
> @miclefebvre
> I specified items from Status to CTL Url in fact definitions, the card in the notification appears like so
> ![image](https://user-images.githubusercontent.com/35904881/97749646-9e738500-1aac-11eb-9d06-8f3080050f9d.png)

You had any success with Mentions?

**Comment by Giusti** [comment]
@tayyab23 
So did you find a solution / workaround how mentions work for this plugin?
I tried using `<at>name<at>` or `<at>group<at>` but the info is just displayed as normal text

**Comment by emmaroberts-nbs** [comment]
This feature would be fantastic. Has anyone made any further progress on this?

**Comment by ljackiewicz** [comment]
I tried to check out what is needed to be done to implement this feature.

Correct me if I am wrong, but if I understand it correctly, the plugin currently only supports the legacy MessageCard format, and in reference to information from a documentation link sent by @tayyab23, mention support is only available for the newer AdaptiveCard format (since version 1.2).

So in this case, there is a need to change/extend the card format first, before an actual implementing of mention support?

**Comment by Giusti** [comment]
@ljackiewicz Yes I thing you are correct, with the current card format mentions do not seem to be possible atleast I havent found a way to get it working

**Comment by leonardobsjr** [comment]
Yeah @ljackiewicz that seems to be the case

**Comment by Giusti** [comment]
@ljackiewicz I was also kind of confused by @tayyab23 comment, because it sounds like he got it working but not sure, he never answered

> I haven't been able to make it work using the existing factDefinitions with at tag around a user like so <at>John Doe</at>

> Edit: Looks like the reason why it didn't work is because I needed the user id wrapped around a mention entity

**Comment by leonardobsjr** [comment]
> @ljackiewicz I was also kind of confused by @tayyab23 comment, because it sounds like he got it working but not sure, he never answered
> 
> > I haven't been able to make it work using the existing factDefinitions with at tag around a user like so `<at>John Doe</at>`
> 
> > Edit: Looks like the reason why it didn't work is because I needed the user id wrapped around a mention entity

I did wrap - the user name was linked, but the user was never notified. Verifying with devtools I saw that the msteams link generation and the one generated from the cards were different - no idea why

**Comment by Giusti** [comment]
@leonardobsjr mh sad, may I ask what exactly you did by wrapping it ?

**Comment by leonardobsjr** [comment]
> @leonardobsjr mh sad, may I ask what exactly you did by wrapping it ?

Below is my full testing json:
`{
    "type":"message",
    "attachments":[
       {
          "contentType":"application/vnd.microsoft.card.adaptive",
          "contentUrl":null,
          "content":{
                "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
                "type": "AdaptiveCard",
                "version": "1.2",
                "body": [
                    {
                        "type": "TextBlock",
                        "id": "MentionTextBlock",
                        "text": "Mention_Test",
                        "weight": "Bolder",
                        "size": "Medium"
                    },
                    {
                        "type": "ColumnSet",
                        "id": "MentionColumnSet",
                        "columns": [
                            {
                                "type": "Column",
                                "id": "MentionColumn",
                                "width": "stretch",
                                "items": [
                                    {
                                        "type": "TextBlock",
                                        "id": "MentionTextBlock2",
                                        "spacing": "None",
                                        "text": "<at>Testing</at>",
                                        "isSubtle": true,
                                        "wrap": true
                                    }
                                ]
                            }
                        ]
                    }
                ],
                "msteams": {
                    "entities": [
                        {
                            "type": "mention",
                            "text": "<at>Testing</at>",
                            "mentioned": {
                                "id": "29:orgid:8a03b5a0-370c-4ebf-a9fe-87fc2d38e2e4",
                                "name": "Testing"
                            }
                        }
                    ]
                }
            }
          }
    ]
 }`
 
 This actually generates the link, but it's kinda of a 'fake link' because it's different from what msteams generates and doesn't notify the user

**Comment by damianszczepanik** [comment]
Guys,
Are aware of that migrating to new adaptive cards means that you need to configure all jobs again? This is because card structure is much different from current one and making sure that migration goes smoothly is quite expensive and risky?
All hooks will stop working until jobs will be reconfigured again

**Comment by leonardobsjr** [comment]
> Guys,
> Are aware of that migrating to new adaptive cards means that you need to configure all jobs again? This is because card structure is much different from current one and making sure that migration goes smoothly is quite expensive and risky?
> All hooks will stop working until jobs will be reconfigured again

I do, but having the ability to mention someone opens so many possibilities that I believe it’s worth it. Maybe this can be done on a fork of the plug-in tho?

**Comment by damianszczepanik** [comment]
You do but how about 13k installations https://plugins.jenkins.io/Office-365-Connector/ and a few times more projects that need to be updated

**Comment by Giusti** [comment]
@damianszczepanik how about adding the adaptive card option instead of changing the existing code? so you could use a different command to use the new adaptive cards includion mentions while the old one would still continue to work?

**Comment by damianszczepanik** [comment]
It is possible, it costs as well :)

**Comment by LukaSvastVolue** [comment]
any progress here, we will really benefit from this feature 
@bogomilkruzicvolue

**Comment by luke-hill** [comment]
Given MSTeams is one of the largest work based chat programs. Tagging someone in a message sent via webhook should be reasonably high up the priority foodchain.

**Comment by vladi14** [comment]
any update about that feature ? would be much appreciated.

**Comment by Talhashahar** [comment]
any update ?

**Comment by precisioninfinity** [comment]
Following, also looking for this feature. Upvote. Please. Thanks in advance.

**Comment by voblers** [comment]
Keeping this thread alive in 2024.
Any updates?

**Comment by taylorb-ml** [comment]
Bumping this again, would be great to tag users/groups in our Teams messages

**Comment by karelkryda** [comment]
Hi,
since support for Adaptive Cards has recently been added, is it now possible to add this functionality too?

Thanks in advance for the info
